README.md

```diff
@@ -1,179 +1,315 @@
+values - so-called 'edge cases', that these generators are very unlikely to find.
 
-  - Generate test suites that search the space of expected behaviours looking for counterexamples.
+We can motivate this with an example, based on <https://github.com/nick8325/quickcheck/issues/98>:
 
+The [abs](https://hackage.haskell.org/package/base/docs/Prelude.html#v:abs)
+function from Prelude is supposed to return the absolute value of a number. 
 
-## What is this?
-`apropos` specifications are composable stochastic search directives. The search is orchestrated by a SAT solver. This approach to testing can be viewed as Satisfiability Modulo Theories (SMT) where the theories are property tests run against code and satisfiability is whether the tested code behaves in the correct way for every test created by the model.
+Therefore, this function should always hold:
 
-While we may not be able to run every property test that the model can generate we can achieve a probabilistic notion of satisfiability that allows us to say that it is very likely that the tested code exactly meets the specification.
-
-### How to write a specification
-Specifications are encoded as logical models associated with a type.
+```haskell
+absIsAlwaysPositive :: Int -> Bool
+absIsAlwaysPositive n = abs n >= 0
+```
 
-```Haskell
-data IntProp
-  = IsNegative
-  | IsPositive
-  | IsZero
-  | IsLarge
-  | IsSmall
-  | IsMaxBound
-  | IsMinBound
-  deriving stock (Eq, Ord, Enum, Show, Bounded)
+Unfortunately, due to two's complement arithmetic, the negation of
+`minBound :: Int` is not representable as an `Int`. Prelude's `abs` function just
+ gives up and returns `minBound`. Hence:
 
-instance Enumerable IntProp where
-  enumerated :: [IntProp]
-  enumerated = [minBound .. maxBound]
+ ```haskell
+ >>> abs minBound :: Int
+-9223372036854775808
+>>> absIsAlwaysPositive minBound
+False
 ```
-Here is a model for integers. We have partitioned the space of integers into a smaller easier to manage space. This will be a sufficient description to test the example function defined later. It is sufficient since the function should return true/false according to a formula defined in terms of the above propositions. By modelling integers like this we also make sure that we test some specific extremal values in the space.
-
-Some of these propositions don't make sense in combination. For example an integer can't be both positive and negative. To express this we use a logical expression to define the space of valid property sets under this model.
-
-```Haskell
-instance LogicalModel IntProp where
-  logic :: Formula IntProp
-  logic =
-    ExactlyOne [Var IsNegative, Var IsPositive, Var IsZero]
-      :&&: ExactlyOne [Var IsLarge, Var IsSmall]
-      :&&: (Var IsZero :->: Var IsSmall)
-      :&&: (Var IsMaxBound :->: (Var IsLarge :&&: Var IsPositive))
-      :&&: (Var IsMinBound :->: (Var IsLarge :&&: Var IsNegative))
+
+Property testing is unlikely to catch this, though:
```

Comment:
Incorrect. QuickCheck is unlikely to find this.
Validity-based testing finds this easily: https://cs-syd.eu/posts/2020-04-28-genvalidity-improvements

Reply:
Naively, it looks like this is a better approach to the same problem. With apropos you have a much better way of defining the generators seemingly.

Reply:
`genvalidity` will solve the problem *in this specific case*. `apropos` provides a framework for finding edge cases *generally*. In general, randomly-generated test data will miss important cases, and bulding custom generators to solve this case-by-case is tedious.